common/include/pcl/common/impl/centroid.hpp

```diff
@@ -658,6 +660,285 @@ computeMeanAndCovarianceMatrix (const pcl::PointCloud<PointT> &cloud,
   return (computeMeanAndCovarianceMatrix (cloud, indices.indices, covariance_matrix, centroid));
 }
 
+template <typename PointT, typename Scalar> inline unsigned int
+computeCentroidAndOBB (const pcl::PointCloud<PointT> &cloud,
+  Eigen::Matrix<Scalar, 3, 1> &centroid,
+  Eigen::Matrix<Scalar, 3, 1> &obb_center,
+  Eigen::Matrix<Scalar, 3, 1> &obb_dimensions,
+  Eigen::Matrix<Scalar, 3, 3> &obb_rotational_matrix)
+{
+  Eigen::Matrix<Scalar, 3, 3> covariance_matrix;
+  Eigen::Matrix<Scalar, 4, 1> centroid4;
+  unsigned int point_count= computeMeanAndCovarianceMatrix(cloud, covariance_matrix, centroid4);
+  if (!point_count)
+    return (0);
+  centroid(0) = centroid4(0);
+  centroid(1) = centroid4(1);
+  centroid(2) = centroid4(2);
+
+  Eigen::SelfAdjointEigenSolver<Eigen::Matrix<Scalar, 3, 3>> evd(covariance_matrix);
+
+  Eigen::Matrix<Scalar, 3, 3> eigenvectors_ = evd.eigenvectors();
+  Eigen::Matrix<Scalar, 3, 1> major_axis;
+  Eigen::Matrix<Scalar, 3, 1> middle_axis;
+  Eigen::Matrix<Scalar, 3, 1> minor_axis;
+
+  minor_axis = eigenvectors_.col(0);//the eigenvectors do not need to be normalized (they are already)
+  middle_axis = eigenvectors_.col(1);
+  // Enforce right hand rule
+  major_axis = middle_axis.cross(minor_axis);
+
+  obb_rotational_matrix <<
+    major_axis(0), middle_axis(0), minor_axis(0),
+    major_axis(1), middle_axis(1), minor_axis(1),
+    major_axis(2), middle_axis(2), minor_axis(2);
+  //obb_rotational_matrix.col(0)==major_axis
+  //obb_rotational_matrix.col(1)==middle_axis
+  //obb_rotational_matrix.col(2)==minor_axis
+
+  //Trasforming the point cloud in the (Centroid, ma-mi-mi_axis) reference
+  //with homogeneous matrix
+  //[R^t  , -R^t*Centroid ]
+  //[0    , 1             ]
+  Eigen::Matrix<Scalar, 4, 4> transform = Eigen::Matrix<Scalar, 4, 4>::Identity();
+  transform.topLeftCorner(3, 3) = obb_rotational_matrix.transpose();
+  transform.topRightCorner(3, 1) =-transform.topLeftCorner(3, 3)*centroid;
+
+  //when Scalar==double on a Windows 10 machine and MSVS:
+  //if you substitute the following Scalars with floats you get a 20% worse processing time, if with 2 PointT 55% worse
+  Scalar obb_min_pointx, obb_min_pointy, obb_min_pointz;
+  Scalar obb_max_pointx, obb_max_pointy, obb_max_pointz;
+  obb_min_pointx = obb_min_pointy = obb_min_pointz = std::numeric_limits<Scalar>::max();
+  obb_max_pointx = obb_max_pointy = obb_max_pointz = std::numeric_limits<Scalar>::min();
+
+  if (cloud.is_dense)
+  {
+    auto point = cloud[0];
```

Comment:
Why do you process this first point outside the for loop?

Else you could have a for-range loop, ie:

```
for(const auto& point : cloud)
{
...
}
```
couldn't you?

anyways, you missed a few .points in the for loops 😄

Reply:
Same for the non-dense branch.

Reply:
for the sane reason I have split the non dense for loop in two parts: to set min max finite so to have

      if (P(0) < obb_min_pointx)
        obb_min_pointx = P(0);
      else if (P(0) > obb_max_pointx)
        obb_max_pointx = P(0); 

instead of

      if (P(0) < obb_min_pointx)
        obb_min_pointx = P(0);
      if (P(0) > obb_max_pointx)
        obb_max_pointx = P(0); 

do you mean cloud.points.size() ??
that is a different thing
I checked there is an inline efficient operator[]
but for this ?

Reply:
Ahh. Make sense.

Yes, I pretty sure you can and should use cloud.size(). The .points should have been a private member, as it exposes the internal container type. Which makes it unchangeable atm, but eventually one could.

Reply:
I guess you save lots of compares  if the first point is in the right "side" and then the next just grows or shrink. Worst case you would have equally as many compares btween the two solutions?

Reply:
one could make difficult statistical analysis on this based on the statistical cloud distribution. for a uniform density the probability would inolv some sort of geometric progression. the best case you save more is when the cloud is a surface with indexes decreasing progressively in x,y,z, the worst the opposite. one could think of a dynamical object that changes the code of comparison based on the cloud progression but let's leave it. I just measured with a counter with one of my main samples and it gives this on 11 segments

saved 0.000258977
saved 0.00359994
saved 0.00205254
saved 0.00217213
saved 0.0025008
saved 0.0554903
saved 0.00217482
saved 0.00121175
saved 0.0190972
saved 0.00453767
saved 0.00445811

Loaded cloud width*height = 2205157 w,h= 2205157 , 1
cloud is_dense= 1
cloud isOrganized= 0

so you can see from 0.02% to 5.5% comparisons avoided, 0.9% in average on this cloud, it can change a lot from cloud to cloud